Automattic/woocommerce-subscriptions-core
# Enable customers to purchase a subscription using WCPay and HPOS tables

Fixes #257 

## Description

This PR includes the changes necessary to allow customers to purchase a subscription using WC Payments + WC Subscriptions with HPOS tables enabled. 

There were 3 separate issues that I was able to troubleshoot and fix as part of this PR. 

1. With HPOS tables enabled purchasing a subscription was leaving the subscription as `Pending` status.
    -  The `WC_Subscriptions_Order::maybe_record_subscription_payment()` function is responsible for activating a subscription after the order is successfully paid. It was not working as expected. 
    - **_Cause:_** The `wcs_order_contains_subscription()` check at the top of the function was causing the function to exit early even though the order did contain a subscription with the parent order relationship. After troubleshooting I discovered it was because of a different in `wc_get_order()` status args. Fixed with [16adf67](https://github.com/Automattic/woocommerce-subscriptions-core/pull/258/commits/16adf67ee1d8fbb8f8566e374e0000b53edc11c8). 
    - Further reading: https://github.com/Automattic/woocommerce-subscriptions-core/pull/258#discussion_r1019899441
2. When a subscription was updated (`$subscription->save()`, after the original create), some key subscription properties (like `_requires_manual_renewal`) weren't being saved to the database (it remained as `true` for example even after the subscriptions payment method was set to WC Payments). 
     - **_Cause:_** In HPOS environments WC orders no longer store key props in metadata. However, Subscriptions still has key data stored in meta (`_requires_manual_renewal`, all the subscription dates, suspension counts etc). Because WC no longer need to store order data in metadata, there isn't a similar function in the order datastore we could override to save these fields for subscriptions. We had a `update_order_meta ` function, however in the Order datastore that only gets called when a order is created, not when it is updated. 
     - **_Fix:_** I've introduced the `persist_updates()` function to now save these subscription props to the database on update. The existing `update_order_meta()` function can still be used to save the subscription props on create.
3. When reading a subscription from the database key subscription props (`_requires_manual_renewal`, all the subscription dates, suspension counts etc) need to be set via the setter (eg `$subscription->set_requires_manual_renewal()`), they aren't read as meta data even though that is where they are stored (meta table). When reading the subscription we call `set_subscription_props()` to set these properties on load, however, there was a bug in this function when it was copied from the CPT datastore equivalent. The `get_post_meta()` became `$subscription->get_meta()` however that isn't a legitimate mapping in this case because these props aren't read into the subscription metadata (they are internal meta keys). To fix this, I updated `set_subscription_props()` to read the subscription's raw meta and pull these internal meta values out of there instead. Fetching the subscription's raw meta does return these internal meta keys. 
4. With database syncing turn off (only HPOS tables), subscriptions were being saved in the tables with a pending status. After a lot of debugging I discovered that this was happening because WC when they save orders and subscriptions in the database call the `$subscription->get_status()` function. The `WC_Subscription::get_status()` function use to pull the subscription status directly from the database and if it is `draft` or `auto-draft` would return pending. Because the post placeholds have a draft status, this would mean calling `$subscription->get_status()` on a HPOS environment with no sycning would essentially always return a pending status. I fixed this in [0989072](https://github.com/Automattic/woocommerce-subscriptions-core/pull/258/commits/09890725776c3ed2f3ad87db019c10d499a3de18)

## How to test this PR

> **Note**
> _**Tip**_ - _test with a non-monthly subscription as monthly is the default period.  eg test a every 4 day subscription or every 2 weeks, 3 years etc._

1. Enable WC Payments + WC Subscriptions + this branch of subscriptions-core.
2.  Enable HPOS tables ([settings](https://user-images.githubusercontent.com/8490476/201551959-1c87e075-d190-4401-93d4-ded8e3ebe921.png)).
3. Purchase a subscription using WC Payments. 
5. Note the following things:
     1. The related subscriptions now appear on the order received page.
     2. View the my account subscription page (on `trunk` you'll need to access the page via the URL - https://example/my-account/view-subscription/400/)
     3. Compare the data in the database. 

| `trunk`  | This branch |
| ------------- | ------------- |
|  <img width="850" alt="Screen Shot 2022-11-14 at 10 44 10 am" src="https://user-images.githubusercontent.com/8490476/201553787-051e7393-eda1-4dc7-b998-8b600cb12a90.png"> |  <img width="850" alt="Screen Shot 2022-11-14 at 10 16 41 am" src="https://user-images.githubusercontent.com/8490476/201552264-90f27a25-da21-4c3a-b522-ab15440d1281.png"> |

| `trunk`  | This branch |
| ------------- | ------------- |
| <img width="864" alt="Screen Shot 2022-11-14 at 10 46 20 am" src="https://user-images.githubusercontent.com/8490476/201553969-11ef6a5a-5de3-4652-bc95-833b163448df.png">  | <img width="880" alt="Screen Shot 2022-11-14 at 10 22 31 am" src="https://user-images.githubusercontent.com/8490476/201552674-4ae365fe-c7b5-4f76-a7ba-63313c18a066.png"> |

## Product impact
